### PR TITLE
Add basic desktop linux checks

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Adds doctor validation checks for gcc and gtk for linux desktop:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v1.5.9-pre.62, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Linux toolchain - develop for Linux desktop applications
[!] Android Studio (version 3.3)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] IntelliJ IDEA Ultimate Edition (version 2018.2)
[!] IntelliJ IDEA Ultimate Edition (version 2018.3)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
[!] VS Code (version 1.27.1)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[✓] Connected device (1 available)
```

With -v
```
[✓] Linux toolchain - develop for Linux desktop applications
    • clang++ 4.0.1
    • GNU Make 4.1
```

https://github.com/flutter/flutter/issues/31368

### CODE REVIEW

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class LinuxDoctorValidator extends DoctorValidator {
+  const LinuxDoctorValidator() : super('Linux toolchain - develop for Linux desktop applications');
+
+  @override
+  Future<ValidationResult> validate() async {
+    ValidationType validationType = ValidationType.installed;
+    final List<ValidationMessage> messages = <ValidationMessage>[];
+    /// Check for some version of GCC.
+    final ProcessResult gccResult = await processManager.run(const <String>[
+      'g++',
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: The choice of GCC vs clang is specific to the FDE example (and might be something we want to change before templatizing); it doesn't seem like we should have `doctor` check for things that aren't required by `flutter_tool`.

How about checking for `make`, since that's what the tooling actually expects at this point?

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    if (gccResult.exitCode != 0) {
+      validationType = ValidationType.missing;
+      messages.add(ValidationMessage.error('GCC executable missing'));
+    } else {
+      messages.add(ValidationMessage(gccResult.stdout));
+    }
+
+    /// Check for the correct version of gtk.
+    final ProcessResult gtkResult = await processManager.run(const <String>[
+      'pkg-config',
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: We'd need to make sure pkg-config was present first. But as with GCC, I don't think we should actually be requiring that. The shell does have a runtime dependency on GTK, but that could be satisfied without pkg-config being present.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/linux/linux_doctor.dart]: The doctor could check make and then for both clang/gcc as a hint? Longer term, I would like to break out each doctor validation check to make it platform/template specific (see go/flutter-tool-extensions).

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/linux/linux_doctor.dart]: These need not be required, but it could be a hint. Is there a better method to detect whether gtk is there at runtime?

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/linux/linux_doctor.dart]: I'm not familiar yet with the gradations of checks that the doctor support, but an error for not having make and a warning for not having gcc or clang definitely seems reasonable as a starting point. (Substituting whatever the right terminology is here for 'error' and 'warning')

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/linux/linux_doctor.dart]: To be clear, when I say it's a runtime dependency I mean that *end users* of the built application need GTK; it's not a developer-specific thing. That's why I don't think `doctor` should check for it.

> Is there a better method to detect whether gtk is there

It's Linux, so as always the answer seems to be "it depends on your distribution!"

Honestly I think that without a bunch of fallback logic and iteration with different distros we're likely to create a lot of false-positives (i.e., whatever we use to check of the presence of GTK is going to be missing more often than GTK), and that we're better off not doing this for now, and seeing if there's really a need for a GTK check in practice based on feedback.

**Review by stuartmorgan-g — COMMENTED** [review]
Sorry, I totally forget about this PR :(

**Comment by jonahwilliams** [comment]
Closing this for now until I have some more time

**Comment by stuartmorgan** [comment]
The template switched to Clang, which I intend to be a permanent change, so when this gets picked up again the initial version that covers everything needed for the simple example app should be:
- check for `make`
- check for `clang++` (if we want to be fancy, version 3.4+, since we use C++14)

**Comment by jonahwilliams** [comment]
This is back on my radar

**Comment by jonahwilliams** [comment]
Updated to check for clang++ greater than version 3.4 and make

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import 'package:flutter_tools/src/base/version.dart';
+
+import '../base/io.dart';
+import '../base/process_manager.dart';
+import '../doctor.dart';
+
+/// A validator that checks for GCC and the correct gtk version.
+class LinuxDoctorValidator extends DoctorValidator {
+  LinuxDoctorValidator() : super('Linux toolchain - develop for Linux desktop applications');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: s/desktop applications/desktop/, since you're developing applications for Linux desktop, not developing something for Linux desktop applications.

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  /// The minimum version of clang supported.
+  final Version minimumClangVersion = Version(3, 4, 0);
+
+  @override
+  Future<ValidationResult> validate() async {
+    ValidationType validationType = ValidationType.installed;
+    final List<ValidationMessage> messages = <ValidationMessage>[];
+    /// Check for some version of Clang.
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: s/some/minimum/

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      validationType = ValidationType.missing;
+      messages.add(ValidationMessage.error('clang++ is not installed.'));
+    } else {
+      final String firstLine = clangResult.stdout.split('\n').first.trim();
+      final String versionString = RegExp(r'[0-9]+\.[0-9]+\.[0-9]+').firstMatch(firstLine).group(0);
+      final Version version = Version.parse(versionString);
+      if (version >= minimumClangVersion) {
+        messages.add(ValidationMessage('clang++ $version'));
+      } else {
+        messages.add(ValidationMessage.error('clang++ $version is below minimum version of $minimumClangVersion'));
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: This should set `ValidationType.partial`.

File: packages/flutter_tools/test/linux/linux_doctor_test.dart [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      expect(result.type, ValidationType.installed);
+      expect(result.messages, <ValidationMessage>[
+        ValidationMessage('clang++ 4.0.1'),
+        ValidationMessage('GNU Make 4.1'),
+      ]);
+    }, overrides: <Type, Generator>{
+      ProcessManager: () => processManager,
+    });
+
+    testUsingContext('Returns partial validation when clang and make are not availible', () async {
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/linux/linux_doctor_test.dart]: s/partial/missing/

File: packages/flutter_tools/test/linux/linux_doctor_test.dart [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      final ValidationResult result = await linuxDoctorValidator.validate();
+      expect(result.type, ValidationType.installed);
+      expect(result.messages, <ValidationMessage>[
+        ValidationMessage('clang++ 4.0.1'),
+        ValidationMessage('GNU Make 4.1'),
+      ]);
+    }, overrides: <Type, Generator>{
+      ProcessManager: () => processManager,
+    });
+
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/linux/linux_doctor_test.dart]: Please add a test that having a version of clang that's below the minimum gives the expected (partial) result ;)

File: packages/flutter_tools/lib/src/linux/linux_doctor.dart [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    }
+
+    /// Check for the correct version of make
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/linux/linux_doctor.dart]: s/correct version of make/make./

(I assume we'll have to tighten this check over time based on experience in the wild, but for now the comment should describe what's actually being checked.)

File: packages/flutter_tools/test/linux/linux_doctor_test.dart [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    setUp(() {
+      processManager = MockProcessManager();
+      linuxDoctorValidator = LinuxDoctorValidator();
+    });
+
+    testUsingContext('Returns full validation when clang++ and make are availibe', () async {
+      when(processManager.run(<String>['clang++', '--version'])).thenAnswer((_) async {
+        return FakeProcessResult(
+          stdout: 'clang version 4.0.1-10 (tags/RELEASE_401/final)\junk',
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/linux/linux_doctor_test.dart]: Was the `\junk` supposed to be `\njunk`?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/test/linux/linux_doctor_test.dart [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+      expect(result.type, ValidationType.partial);
+      expect(result.messages, <ValidationMessage>[
+        ValidationMessage.error('clang++ 2.0.1 is below minimum version of 3.4.0'),
+        ValidationMessage('GNU Make 4.1'),
+      ]);
+    }, overrides: <Type, Generator>{
+      ProcessManager: () => processManager,
+    });
+
+    testUsingContext('Returns missing validation when clang and make are not availible', () async {
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/linux/linux_doctor_test.dart]: It just occurred to me that this would be better as two tests, one where clang is good and make is missing, and one where make is good and clang is missing, to make sure each path correctly sets `missing`; as is a bug in one path could be masked by correct code in the other.

**Review by stuartmorgan-g — APPROVED** [review]
One more test note, but LGTM

**Comment by jonahwilliams** [comment]
cc @dnfield, do you know why this change would be failing the release_smoke_tests ? I don't know how to run it locally to debug, but this should have no effect on an Android build

**Comment by dnfield** [comment]
It looks like Gradle is failing to build in release mode with these changes.

The command that's failing is `../../../bin/flutter build appbundle --target-platform android-arm,android-arm64` from `dev/integration_tests/release_smoke_test`.  Is that passing locally?

**Comment by jonahwilliams** [comment]
Works fine locally, though I'm also on a mac

**Comment by dnfield** [comment]
Ah.  I think this is the gradle emoji bug.

Can you try replacing:
```
        - ./dev/bots/firebase_testlab.sh
```

with 


```
        - echo "$CIRRUS_CHANGE_MESSAGE" > /tmp/cirrus_change_message.txt
        - echo "$CIRRUS_COMMIT_MESSAGE" > /tmp/cirrus_commit_message.txt
        - export CIRRUS_CHANGE_MESSAGE=""
        - export CIRRUS_COMMIT_MESSAGE=""
        - ./dev/bots/firebase_testlab.sh
        - export CIRRUS_CHANGE_MESSAGE=`cat /tmp/cirrus_change_message.txt`
        - export CIRRUS_COMMIT_MESSAGE=`cat /tmp/cirrus_commit_message.txt`
```

in cirrus.yml?

**Comment by jonahwilliams** [comment]
Didn't work, can we submit on red for this case after I roll back the cirrus config?

**Comment by tvolkert** [comment]
Looks like this is causing the following bot failures:

```
[✓] Flutter (Channel master, v1.7.6-pre.19, on Linux, locale en_US.UTF-8)
    • Flutter version 1.7.6-pre.19 at /tmp/flutter sdk
    • Framework revision f221ad1f92 (3 minutes ago), 2019-06-19 15:51:21 -0700
    • Engine revision 54f88ab5da
    • Dart version 2.4.0
Invalid argument(s): Cannot find executable for clang++.
#0      _getExecutable (package:process/src/interface/local_process_manager.dart:113:5)
#1      LocalProcessManager.run (package:process/src/interface/local_process_manager.dart:62:7)
#2      LinuxDoctorValidator.validate (package:flutter_tools/src/linux/linux_doctor.dart:22:60)
<asynchronous suspension>
#3      Doctor.startValidatorTasks (package:flutter_tools/src/doctor.dart:152:52)
#4      Doctor.diagnose (package:flutter_tools/src/doctor.dart:223:41)
#5      _AsyncAwaitCompleter.start (dart:async-patch/async_patch.dart:49:6)
#6      Doctor.diagnose (package:flutter_tools/src/doctor.dart:213:24)
#7      DoctorCommand.runCommand (package:flutter_tools/src/commands/doctor.dart:48:39)
```

**Comment by jonahwilliams** [comment]
Not sure why CI didn't catch this - possibly some interaction between the PRs I landed today
